### DISCUSSION TITLE
Syntax highlighting and formatting

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#16706] [Features and ideas]
In the current backoffice we have a `<umb-code-snippet>` component, which doesn't really highlight or format the code snippet.
When I added it back in the days I considered [Prism](https://prismjs.com/) or [highlightjs](https://highlightjs.org/), but we didn't want to include another library at that time.

We could probably have had a readonly instance of Ace Editor.

In the new backoffice the code editor is replaced by [Monaco editor](https://microsoft.github.io/monaco-editor/)

I wonder if we can use the as code formatting or any other lightweight library for code formatting and syntax highlighting? I think it could be useful in packages showing a code snippet or maybe some help to more technical content editors.

https://stackoverflow.com/questions/52700307/how-to-use-monaco-editor-for-syntax-highlighting

**Comment by bjarnef** [comment]
I came across https://shiki.matsu.io/ for syntax highlighting which has integration with Monaco editor allowing  to use Shiki's syntax highlighting engine to highlight Monaco: https://shiki.matsu.io/packages/monaco

@iOvergaard @nielslyngsoe some thoughts, which may be relevant regarding themes in backoffice.

**Comment by bjarnef** [comment]
I think we can use the `colorizeElement` function from monaco editor to handle this.
https://stackoverflow.com/a/53162251

There's a demo of it here:
https://microsoft.github.io/monaco-editor/playground.html?source=v0.52.2#example-creating-the-editor-syntax-highlighting-for-html-elements

@iOvergaard @leekelleher @nielslyngsoe

**Reply by bjarnef** [reply]
It seems there may be an bug using JSON, but a workaround: https://github.com/microsoft/monaco-editor/issues/3105#issuecomment-2562693504

**Reply by nielslyngsoe** [reply]
Hi @bjarnef, I agree it's easier to read with colors, but also a feature that could be heavy to pull in. But I'm happy to read that it seems we can support this with a library we already have in the project. So very nice, looking ofrward to see how it tursn out.

**Reply by bjarnef** [reply]
@nielslyngsoe if you or someone else in team has time, one can try out this PR https://github.com/umbraco/Umbraco-CMS/pull/19813

I got some errors if I tried to export `jsonDefaults` and `setupMode` as part of the monaco editor bundle.

**Comment by bjarnef** [comment]
It can possible be used to highlight the CSS in RTE styles preview as well by combining the selector and the properties and possible lock the CSS selector.
https://dev.to/pranomvignesh/restrict-editable-area-in-monaco-editor-4hac

**Comment by bjarnef** [comment]
Another place where it would be useful is in the webhook delivery logs in Umbraco.

E.g. some similar to this shown in Adyen webhooks for JSON response:
https://docs.adyen.com/development-resources/webhooks/

**Comment by bjarnef** [comment]
I also found this https://andreruffert.github.io/syntax-highlight-element/ which use [Prism](https://github.com/PrismJS/prism)'s tokenizer and the [CSS Custom Highlight API](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Custom_Highlight_API) for syntax highlighting.
